server/src/main/java/org/elasticsearch/index/mapper/IgnoredValuesFieldMapper.java

```diff
@@ -0,0 +1,188 @@
+
+    static final NodeFeature TRACK_IGNORED_VALUES = new NodeFeature("mapper.track_ignored_values");
+
+    /*
+     * Container for the ignored field data:
+     *  - the full name
+     *  - the offset in the full name indicating the end of the substring matching
+     *    the full name of the parent field
+     *  - the value, encoded as a byte array
+     */
+    public record Values(String name, int parentOffset, BytesRef value) {}
+
+    static final class IgnoredValuesFieldMapperType extends StringFieldType {
+
+        private static final IgnoredValuesFieldMapperType INSTANCE = new IgnoredValuesFieldMapperType();
+
+        private IgnoredValuesFieldMapperType() {
+            super(NAME, false, true, false, TextSearchInfo.NONE, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return NAME;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
+            return new StoredValueFetcher(context.lookup(), NAME);
+        }
+    }
+
+    private IgnoredValuesFieldMapper() {
+        super(IgnoredValuesFieldMapperType.INSTANCE);
+    }
+
+    @Override
+    protected String contentType() {
+        return NAME;
+    }
+
+    @Override
+    public void postParse(DocumentParserContext context) {
+        for (Values values : context.getIgnoredFieldValues()) {
+            context.doc().add(new StoredField(NAME, encode(values)));
+        }
+    }
+
+    static byte[] encode(Values values) {
+        assert values.parentOffset < PARENT_OFFSET_IN_NAME_OFFSET;
+        assert values.parentOffset * (long) PARENT_OFFSET_IN_NAME_OFFSET < Integer.MAX_VALUE;
+
+        byte[] nameBytes = values.name.getBytes(StandardCharsets.UTF_8);
+        byte[] bytes = new byte[4 + nameBytes.length + values.value.length];
+        ByteUtils.writeIntLE(values.name.length() + PARENT_OFFSET_IN_NAME_OFFSET * values.parentOffset, bytes, 0);
+        System.arraycopy(nameBytes, 0, bytes, 4, nameBytes.length);
+        System.arraycopy(values.value.bytes, values.value.offset, bytes, 4 + nameBytes.length, values.value.length);
+        return bytes;
+    }
+
+    static Values decode(Object field) {
+        byte[] bytes = ((BytesRef) field).bytes;
+        int encodedSize = ByteUtils.readIntLE(bytes, 0);
+        int nameSize = encodedSize % PARENT_OFFSET_IN_NAME_OFFSET;
+        int parentOffset = encodedSize / PARENT_OFFSET_IN_NAME_OFFSET;
+        String name = new String(bytes, 4, nameSize, StandardCharsets.UTF_8);
+        BytesRef value = new BytesRef(bytes, 4 + nameSize, bytes.length - nameSize - 4);
+        return new Values(name, parentOffset, value);
+    }
+
+    @Override
+    public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
+        return new SyntheticLoader();
+    }
+
+    public static class SyntheticLoader implements SourceLoader.SyntheticFieldLoader {
```

Comment:
I wonder whether this logic can be simplified. This currently inlines the ignored field values into the _source being constructed, but the logic adds complexity here and in `ObjectMapper`. 

An alternative implementation would be not to inline the ignored field values here. But just synthesize the _source as if there are no ignored field values. Then create a separate source that only contains the ignored field values and merge that with other source. 

I think we can maybe do this in the  `SourceLoaderSynthetic.SyntheticLeaf#source(...)` method. From the leaf reader we can figure out whether there is the current document has an ignored field value stored field. If that isn't the case, then the logic that exists today just gets executed. Otherwise, we load the synthetic source, then load the field values form `_ignored_values` and turn that into a map. Merge that (using something like `XContentHelper#merge(...)`) with the synthetic source and return that.  I think this could result in less complex logic? This would be more expensive, but only if field values end up being ignored.

Reply:
I also thought of the merge approach, but couldn't figure how to get it to work. In your suggestion, how can I get the `Map<String, Object>` merged into the `XContentBuilder`? I checked other uses of `XContentHelper#merge`, they mostly apply to mappings.

Reply:
Updated based on your suggestion, moving ignored value fetching and decoding to `SourceLoader`.

Reply:
👍 thanks, looks simpler now.